### PR TITLE
KAFKA-17104: fix InvalidMessageCrcRecordsPerSec is not updated in validating Legac…

### PR DESCRIPTION
**Pull request by TaiJuWu**
…yRecord

*More detailed description of your change,

*Summary of testing strategy (including rationale)
Uint test

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/test/scala/unit/kafka/log/LogValidatorTest.scala [hunk 1]
```diff
@@ -687,6 +687,57 @@ class LogValidatorTest {
     verifyRecordValidationStats(validatedResults.recordValidationStats, numConvertedRecords = 0, records,
       compressed = true)
   }
+  @Test
```
Comment (chia7712) [inline core/src/test/scala/unit/kafka/log/LogValidatorTest.scala]: Could you please use `ValueSource` or `CsvSource`?

**Review by chia7712 — COMMENTED** [review]
@TaiJuWu could you please update docs ?https://github.com/apache/kafka/blob/ec9dabf86fd0ddd006cf26cd5d4dc967fdd6826f/clients/src/main/java/org/apache/kafka/common/record/LegacyRecord.java#L127

File: core/src/test/scala/unit/kafka/log/LogValidatorTest.scala (hunk 1 above)
Comment (TaiJuWu) [reply core/src/test/scala/unit/kafka/log/LogValidatorTest.scala]: Hi @chia7712 thanks for review.
I found valueSource and CsvSource is not an easy way.
How about MehodSource ?

Comment (chia7712) [reply core/src/test/scala/unit/kafka/log/LogValidatorTest.scala]:
```scala
  @ParameterizedTest
  @CsvSource(Array("0,gzip", "1,gzip", "0,lz4", "1,lz4", "0,snappy", "1,snappy"))
  def testInvalidChecksum(code: Byte, compression: String): Unit = {
    checkInvalidChecksum(code, CompressionType.forName(compression))
  }
```

how about this?

Comment (TaiJuWu) [reply core/src/test/scala/unit/kafka/log/LogValidatorTest.scala]: WoW, this is so great. Thanks!
I just want to build class with `CsvSource` and not think this method...

**Comment by TaiJuWu** [comment]
> @TaiJuWu could you please update docs ?
> 
> https://github.com/apache/kafka/blob/ec9dabf86fd0ddd006cf26cd5d4dc967fdd6826f/clients/src/main/java/org/apache/kafka/common/record/LegacyRecord.java#L127

Done. Thanks for your check.

**Comment by chia7712** [comment]
@TaiJuWu please rebase code to trigger QA again

**Comment by TaiJuWu** [comment]
> @TaiJuWu please rebase code to trigger QA again

Hi @chia7712 , It seems not CI issue because it appear two times.
I guess number from `CsvSource` can't be converted into `Byte` directly with old scala/java version.
Or maybe there are other reasons.
I will build environment to test it if needed.

Or addess this in https://github.com/apache/kafka/pull/16167 ?
WDYT?

**Comment by chia7712** [comment]
> I guess number from CsvSource can't be converted into Byte directly with old scala/java version.
Or maybe there are other reasons.
I will build environment to test it if needed.

not really. The deadlock is caused by `testLogRecoveryMetrics`, and it will be fixed by #16614

**Comment by TaiJuWu** [comment]
> > I guess number from CsvSource can't be converted into Byte directly with old scala/java version.
> > Or maybe there are other reasons.
> > I will build environment to test it if needed.
> 
> not really. The deadlock is caused by `testLogRecoveryMetrics`, and it will be fixed by #16614

OK. Thanks your review and information.

File: storage/src/main/java/org/apache/kafka/storage/internals/log/LogValidator.java [hunk 2]
```diff
@@ -574,7 +574,7 @@ private static Optional<ApiRecordError> validateRecord(RecordBatch batch,
         if (batch.magic() <= RecordBatch.MAGIC_VALUE_V1 && batch.isCompressed()) {
             try {
                 record.ensureValid();
-            } catch (InvalidRecordException e) {
```
Comment (chia7712) [inline storage/src/main/java/org/apache/kafka/storage/internals/log/LogValidator.java:577]: In order to keep consistency, we should rethrow  `CorruptRecordException` here. for example:
```java
            } catch (CorruptRecordException e) {
                metricsRecorder.recordInvalidChecksums();
                throw e;
            } catch (InvalidRecordException e) {
                metricsRecorder.recordInvalidChecksums();
                throw new CorruptRecordException(e.getMessage() + " in topic partition " + topicPartition);
            }
```

Comment (TaiJuWu) [reply storage/src/main/java/org/apache/kafka/storage/internals/log/LogValidator.java:577]: Done. Thanks for this suggestion.

**Review by chia7712 — APPROVED** [review]
LGTM

**Pull request merged**